### PR TITLE
Ajax Widget

### PR DESCRIPTION
**Pull request by stphnpt**
### Changelog

``` markdown
### Added
- Added `AjaxBlockService`
```
### Subject

The goal of this PR is to add a very simple widget that will fit most of the needs.

This widget needs an Url (can be external) which is going to be called to
populate dashboard data. The render is fully customisable  with AdminLTE referenced colors,
the widget provides some default values and a choice of three different templates as
depicted below.
capture d ecran 2016-07-04 a 23 31 54
Widgets can be updated by a click on it.
capture d ecran 2016-07-05 a 00 19 46

See #7 #8

a6f5bf3c-3c44-11e5-91ea-151af7c5eecc
### To do
- [ ] Update the documentation
- [ ] Use a namespace for javascript
- [ ] add a setting for an auto-refresh timeout
- [ ] Choices needs to be flipped for symfony 2.8 in AjaxBlockService

### CODE REVIEW

File: Resources/assets_src/src/js/ajax-block.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+	const getAllBlocks = el => Array.from(el.querySelectorAll('.sonata-dashboard-ajax-block'))
+	const populateBlock = (block, html) => block.innerHTML = html
+	const fetchBlock = block => fetch(block.dataset.url)
+	  .then(response => response.text())
+	  .catch(error => `/!\ Failed to load ${block.dataset.url}`)
+	  .then(html => populateBlock(block, html))
+	const fetchAllBlocks = el => Promise.all(getAllBlocks(el).map(fetchBlock))
+
+	document.addEventListener('DOMContentLoaded', () => fetchAllBlocks(document).then(() => {
```
Comment (greg0ire) [inline Resources/assets_src/src/js/ajax-block.js]: Shouldn't lines 14 through 22 be indented?

Comment (stphnpt) [inline Resources/assets_src/src/js/ajax-block.js]: That file has to be completely reworked as ju1ius mentionned in the previous PR #47

File: Block/AjaxBlockService.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        return 'sonata.dashboard.ajax';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getBlockMetadata($code = null)
+    {
+        return new Metadata($this->getName(),
+            ($code ?: $this->getName()), false, 'SonataDashboardBundle', array('class' => 'fa fa-dashboard'));
```
Comment (greg0ire) [inline Block/AjaxBlockService.php]: If you use multiline style, each argument must be on its own line.

File: Resources/public/sonata-dashboard.front.css [hunk 3]
```diff
@@ -79,15 +79,6 @@
...
 
-/*# sourceMappingURL=dashboard.css.map */
-
-{
-"version": 3,
-"mappings": ";AACA,yBAA0B;EACtB,MAAM,EAAE,IAAI;;AAEhB,+BAAgC;EAC5B,KAAK,EAAE,OAAO;;AAGlB,mBAAoB;EAChB,UAAU,EAAE,MAAM;;AAGtB,kCAAmC;EAC/B,UAAU,EAAE,OAAO;;;AAIvB,yBAA0B;EACtB,QAAQ,EAAE,QAAQ;EAClB,UAAU,EAAE,IAAI;;;AAIpB,qCAAsC;EAClC,gBAAgB,EAAE,sBAAsB;;;;AAQ5C,qCAAsC;EAClC,OAAO,EAAE,YAAY;EACrB,OAAO,EAAE,GAAG;EACZ,SAAS,EAAE,IAAI;EACf,KAAK,EAAE,OAAO;;AAElB,6DAA8D;EAC1D,OAAO,EAAE,KAAK;EACd,KAAK,EAAE,IAAI;EACX,gBAAgB,EAAE,OAAmB;;AAEzC,yDAA0D;EACtD,OAAO,EAAE,KAAK;EACd,gBAAgB,EAAE,OAAkB;;;AAGxC,kFAAmF;EAC/E,OAAO,EAAE,KAAK;;AAElB,6EAA8E;EAC1E,OAAO,EAAE,KAAK;;;AAIlB,6BAA8B;EAC1B,WAAW,EAAE,IAAI;EACjB,MAAM,EAAE,kBAAiC;;;;;;AAQ7C,qCAAsC;EAClC,gBAAgB,EAAE,wBAAwB;EAC1C,UAAU,EAAE,IAAI;;;AAIpB,sCAAuC;EACnC,MAAM,EAAE,IAAI;;;AAIhB,+DAAgE;;EAE5D,gBAAgB,EAAE,sBAAsB",
-"sources": ["../dashboard.scss"],
-"names": [],
-"file": "dashboard.css"
-}
```
Comment (greg0ire) [inline Resources/public/sonata-dashboard.front.css]: Isn't that already done in your other PR ? If yes please rebase, it will make the diff more simple.

File: Resources/assets_src/src/js/ajax-block.js (hunk 1 above)
Comment (rande) [inline Resources/assets_src/src/js/ajax-block.js]: is this JS is compliant with IE ?

Comment (Soullivaneuh) [inline Resources/assets_src/src/js/ajax-block.js]: IWhat? :trollface:

**Comment by core23** [comment]
Please wait for https://github.com/sonata-project/SonataBlockBundle/issues/282

File: Block/AjaxBlockService.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            'class' => 'col-md-3 col-sm-6 col-xs-12',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return 'sonata.dashboard.ajax';
```
Comment (core23) [inline Block/AjaxBlockService.php]: Please use the same name as the block service id.

**Comment by core23** [comment]
Please don't forget the translation keys.

File: Block/AjaxBlockService.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setDefaultSettings(OptionsResolverInterface $resolver)
+    {
+        $resolver->setDefaults(array(
+            'template' => 'SonataDashboardBundle:Block:ajax.html.twig',
```
Comment (core23) [inline Block/AjaxBlockService.php]: Templates starts with `block_*.html.twig`

File: Block/AjaxBlockService.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildEditForm(FormMapper $form, BlockInterface $block)
+    {
+        $form->add('settings', 'sonata_type_immutable_array', array(
+            'label' => false,
+            'keys' => array(
+                array('title', 'text', array()),
```
Comment (core23) [inline Block/AjaxBlockService.php]: Please add labels.

File: Tests/Block/AjaxBlockServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+namespace Sonata\DashboardBundle\Tests\Block;
+
+use Sonata\BlockBundle\Block\BlockContext;
+use Sonata\BlockBundle\Model\Block;
+use Sonata\BlockBundle\Tests\Block\Service\FakeTemplating;
+use Sonata\DashboardBundle\Block\AjaxBlockService;
+
+final class AjaxBlockServiceTest extends \PHPUnit_Framework_TestCase
```
Comment (core23) [inline Tests/Block/AjaxBlockServiceTest.php]: Please use `Sonata\BlockBundle\Tests\Block\AbstractBlockServiceTest`.

File: Resources/assets_src/src/js/ajax-block.js (hunk 1 above)
Comment (core23) [inline Resources/assets_src/src/js/ajax-block.js]: Which IEs do we support? Only supported desktop version (IE9+)?

**Comment by stphnpt** [comment]
AFAIK, javascript of ColorPicker in CoreBundle has to be reviewed (KO)

**Comment by OskarStark** [comment]
@stphnpt could you please add the screenshot from #8 to the PR header? thank you!

**Comment by stphnpt** [comment]
@OskarStark done.

File: Resources/assets_src/src/js/ajax-block.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ * file that was distributed with this source code.
+ *
+ */
+
+(function ($) {
+
+    function fetchBlock ($block) {
+        
+    	// weird things happen when $.ajax with empty url
+    	if ($block.data('url') != '') {
```
Comment (core23) [inline Resources/assets_src/src/js/ajax-block.js]: Indention looks wrong

File: Resources/translations/SonataDashboardBundle.en.xliff [hunk 9]
```diff
@@ -214,6 +214,22 @@
                 <source>form.label_template</source>
                 <target>Template</target>
             </trans-unit>
+            <trans-unit id="53" resname="form.label_title">
```
Comment (core23) [inline Resources/translations/SonataDashboardBundle.en.xliff]: Don't forget the other translation files! 

Please use the same `target` translation if you don't know the correct translation.

Comment (core23) [inline Resources/translations/SonataDashboardBundle.en.xliff]: For the German file:

form.label_title: Titel
form.label_url: URL
form.label_icon: Icon
form.label_color: Farbe

Comment (stphnpt) [inline Resources/translations/SonataDashboardBundle.en.xliff]: @core23 done.

File: Block/AjaxBlockService.php [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolverInterface;
+
+/**
+ * @author  Quentin Somazzi <qsomazzi@ekino.com>
+ */
+final class AjaxBlockService extends BaseBlockService
+{
+    /**
+     * @var string[]
```
Comment (core23) [inline Block/AjaxBlockService.php]: Indention looks wrong. Why the linebreak?

Same for the other phpdocs.

File: Resources/views/Block/block_ajax_progress.html.twig [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+                    <div class="progress-bar" style="width: {{ block.settings.progress*100}}%"></div>
+                </div>
+                <span class="progress-description">
+                    {{ block.settings.description }}
+                </span>
+            </div>
+        </div>
+    </div>
+{% endblock block_content %}
+
```
Comment (core23) [inline Resources/views/Block/block_ajax_progress.html.twig]: No extra line breaks here

File: Resources/views/Block/block_ajax_simple.html.twig [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    <div class="overlay-wrapper sonata-dashboard-ajax-block__refresh-btn">
+        <div class="info-box">
+            <span class="info-box-icon bg-{{ block.settings.color }}"><i class="{{ block.settings.icon }}"></i></span>
+            <div class="info-box-content">
+                <span class="info-box-text">{{ block.settings.title }}</span>
+                <span class="info-box-number url-data">-</span>
+            </div>
+        </div>
+    </div>
+{% endblock block_content %}
```
Comment (core23) [inline Resources/views/Block/block_ajax_simple.html.twig]: Please add a linebreak

Comment (Soullivaneuh) [inline Resources/views/Block/block_ajax_simple.html.twig]: linebreak -> new line

File: Resources/assets_src/src/js/ajax-block.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ */
+
+(function ($) {
+
+    function fetchBlock ($block) {
```
Comment (core23) [inline Resources/assets_src/src/js/ajax-block.js]: Please try to use a namespace for javascript.

**Comment by core23** [comment]
Ping @stphnpt

File: Block/AjaxBlockService.php (hunk 10 above)
Comment (stphnpt) [inline Block/AjaxBlockService.php]: Looks wrong but seems ok (styleCI)

**Comment by greg0ire** [comment]
ping @Soullivaneuh for stuck build

**Comment by core23** [comment]
StyleCI @Soullivaneuh  ;)

**Comment by rande** [comment]
Can we add a setting for an auto-refresh timeout ?

File: Block/AjaxBlockService.php [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+                )),
+                array('icon', 'text', array(
+                    'required' => false,
+                    'label' => 'form.label_icon',
+                    'attr' => array('placeholder' => 'fa fa-dashboard'),
+                )),
+                array('color', 'choice', array(
+                     'required' => true,
+                     'label' => 'form.label_color',
+                     'choices' => self::$colors,
```
Comment (core23) [inline Block/AjaxBlockService.php]: `choices` needs to be flipped for symfony 2.8

File: Block/AjaxBlockService.php [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+     */
+    public function buildEditForm(FormMapper $form, BlockInterface $block)
+    {
+        $form->add('settings', 'sonata_type_immutable_array', array(
+            'label' => false,
+            'keys' => array(
+                array('template', 'choice', array(
+                    'required' => true,
+                    'label' => 'form.label_template',
+                    'choices' => self::$templates,
```
Comment (core23) [inline Block/AjaxBlockService.php]: Please flip this for symfony 2.8

File: Tests/Block/AjaxBlockServiceTest.php [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace Sonata\DashboardBundle\Tests\Block;
+
+use Sonata\BlockBundle\Block\BlockContext;
+use Sonata\BlockBundle\Model\Block;
+use Sonata\BlockBundle\Tests\Block\AbstractBlockServiceTest;
+use Sonata\BlockBundle\Tests\Block\Service\FakeTemplating;
+use Sonata\DashboardBundle\Block\AjaxBlockService;
+
+final class AjaxBlockServiceTest extends AbstractBlockServiceTest
```
Comment (core23) [inline Tests/Block/AjaxBlockServiceTest.php]: We don't need final anymore for test classes

File: Tests/Block/AjaxBlockServiceTest.php [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\DashboardBundle\Tests\Block;
+
+use Sonata\BlockBundle\Block\BlockContext;
+use Sonata\BlockBundle\Model\Block;
+use Sonata\BlockBundle\Tests\Block\AbstractBlockServiceTest;
```
Comment (core23) [inline Tests/Block/AjaxBlockServiceTest.php]: This class was moved to the `Test` namespace

**Comment by core23** [comment]
Ping @stphnpt

**Comment by core23** [comment]
Are you still working on this @stphnpt ?

**Comment by stphnpt** [comment]
I'm sorry but for the time being, I don't have time to follow on this work ...

**Comment by core23** [comment]
Very sad. Maybe you could give a short status for this feature? Should it work for now or is there much to do?

**Comment by stphnpt** [comment]
Just a few things to finish but it was fully functional in june last year : "try to use a namespace for javascript", "choices needs to be flipped for symfony 2.8" and "We don't need final anymore on test classes", you might want to do it yourself, no ?

**Comment by OskarStark** [comment]
can you please add this as a todo list in the PR header?

**Comment by stphnpt** [comment]
done !

**Comment by OskarStark** [comment]
@stphnpt can you please rebase, I think travis is green afterwards

File: Block/AjaxBlockService.php [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+use Sonata\BlockBundle\Block\Service\AbstractBlockService;
+use Sonata\BlockBundle\Model\BlockInterface;
+use Sonata\CoreBundle\Model\Metadata;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author  Quentin Somazzi <qsomazzi@ekino.com>
+ */
+class AjaxBlockService extends AbstractBlockService
```
Comment (core23) [inline Block/AjaxBlockService.php]: Please make this final

File: Block/AjaxBlockService.php [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+        'purple' => 'purple_color',
+        'maroon' => 'maroon_color',
+        'black' => 'black_color',
+    );
+
+    /**
+     * @var string[]
+     *               Each template corresponds to the AdminLTE 2.3.3 widgets
+     */
+    public static $templates = array(
```
Comment (core23) [inline Block/AjaxBlockService.php]: Can we make this private?

**Comment by core23** [comment]
Please have a look at the open tasks, so we can merge this @stphnpt

**Comment by core23** [comment]
Not sure if this is the right location for this block. Maybe this should be moved to the BlockBundle?

**Comment by core23** [comment]
Move the stuff to the block bundle
